### DISCUSSION TITLE
Be able to trigger an action on click of (sub-string of) text in a notify

### DISCUSSION THREAD

**Discussion by davetapley** [Textualize/textual#3526] [Ideas]
My use case:

1. User clicks button to start processing job in the background.
2. When job is done I `notify` to let them know it's done.
3. The file is available at URL. 

Right now I can make that URL 'actionable' with a button that `webbrowser.open(url)`'s ,
but I'd _love it_ if they could click the notification.

---

Two possibilities:

Easy: clicking notification triggers function.

Stretch: Be able to define 'link' in text which is clickable
(and maybe 'links' could be more broadly be available for strings in Textual).

**Comment by TomJGooding** [comment]
> and maybe 'links' could be more broadly be available for strings in Textual

I think you can output links with [Rich markup](https://rich.readthedocs.io/en/latest/markup.html#links) which it looks like `notify` supports from the [examples](https://textual.textualize.io/widgets/toast/#example). Sorry not at my computer at the moment to create a test, but give that a try?

**Reply by TomJGooding** [reply]
After some quick experimentation, actually I don't think Rich markup for links does work in Textual. Not sure if it should so I'll do more digging before raising an issue.

However you can achieve this using Textual's [action Links](https://textual.textualize.io/styles/links/) - example below. 

Unfortunately I had to redefine all the styles as otherwise you couldn't make out the link - perhaps another issue.

```python
import webbrowser

from textual.app import App


class ToastLinkApp(App):
    CSS = """
    Toast {
        link-background:;
        link-color: $text;
        link-style: underline;
        link-hover-background: $accent;
        link-hover-color: $text;
        link-hover-style: bold not underline;
    }
    """

    def on_mount(self) -> None:
        self.notify(
            "Don't forget to "
            "[@click=link('https://github.com/Textualize/textual')]star[/] "
            "the repository!",
            timeout=10,
        )

    def action_link(self, href: str) -> None:
        webbrowser.open(href)


if __name__ == "__main__":
    app = ToastLinkApp()
    app.run()
```

**Reply by davetapley** [reply]
Ahhhh 'action links', I knew I've seen a feature like that before, but I couldn't find it and figured I'd dreamed it 😆 

Thanks and especially thanks for the styling. I'll have a play around and open an issue for that if/when I can figure it out.

---


Also I _assume_ Textual actions links are better than Rich text links because the latter have the caveat:
> If your terminal software supports hyperlinks, you will be able to click the word “blog” which will typically open a browser. If your terminal doesn’t support hyperlinks, you will see the text but it won’t be clickable.

But maybe they use the same mechanism under the hood anyway 🤔 

Any insight @willmcgugan ? 🙏🏻

**Reply by TomJGooding** [reply]
> Thanks and especially thanks for the styling. I'll have a play around and open an issue for that if/when I can figure it out.

No problem, my memory of these Textual links was a bit hazy too!

The `Toast` widget used by `notify` doesn't inherit the styling from the base class, so the "default" links styling defined in `Widget` is lost. You can obviously just add it back in like I've done here, but it might be nicer if you didn't have to do that - I might look at a PR.

**Reply by willmcgugan** [reply]
Links in Rich use a terminal protocol, which is fairly limited. Basically an app can't respond to them.

Action links are entirely managed by Textual, so you can do pretty much anything with them.

**Reply by TomJGooding** [reply]
Thanks Will, so just to clarify Rich link markup won't work in Textual? I found a couple of mentions in the docs which suggests they do.

**Reply by willmcgugan** [reply]
I'm surprised that they don't.

**Comment by darrenburns** [comment]
I just merged this PR from @TomJGooding - it was just some missing styling, the links in the notify were always actually clickable, it just wasn't obvious because they blended in with normal text - https://github.com/Textualize/textual/pull/3531

**Reply by TomJGooding** [reply]
Thanks Darren! Just to clarify for posterity, my PR was just a styling fix for **Textual** links. Whether [Rich link markup](https://rich.readthedocs.io/en/latest/markup.html#links) should also work is a different question...
